File: 라이 대회알고리즘/day05. 이분 탐색(Binary Search)/2805_나무 자르기.py
import sys

readline = sys.stdin.readline


def step1(N, M, input):
    left, right, mid = 0, max(input), 0

    while left <= right:
        mid = (left + right) // 2
        answer = 0
        for i in range(0, N):
            answer += max(input[i] - mid, 0)

        # M에 도달했다고 답이 아니라 높이의 최대값을 출력해야함
        # if answer == M:
        #     break
        if answer >= M:
            left = mid+1
        else:
            right = mid-1
    return right


if __name__ == '__main__':
    N, M = map(int, readline().split())

    input = list(map(int, readline().split()))

    print(step1(N, M, input))

File: 라이 대회알고리즘/day05. 이분 탐색(Binary Search)/2805_나무 자르기.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	UintSize = 32 << (^uint(0) >> 32 & 1)
	MaxInt   = 1<<(UintSize-1) - 1 // 1<<31 - 1 or 1<<63 - 1
	MinInt   = -MaxInt - 1
)

func max(numbers ...int) (result int) {
	result = MinInt
	for _, number := range numbers {
		if result < number {
			result = number
		}
	}
	return result
}

func step1(N int, M int, tree []int) (result int) {

	lb, rb := 0, max(tree...)

	for lb <= rb {
		mid := (lb + rb) / 2

		var answer = 0
		for i := range tree {
			answer += max(tree[i]-mid, 0)
		}

		if answer >= M {
			lb = mid + 1
		} else {
			rb = mid - 1
		}
	}
	return rb
}

func main() {

	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)

	var N, M int
	fmt.Fscanln(reader, &N, &M)

	tree := make([]int, N)
	for i := 0; i < N; i++ {
		fmt.Fscanf(reader, "%d ", &tree[i])
	}

	fmt.Fprintln(writer, step1(N, M, tree))
	writer.Flush()
}
